spring-integration-core/src/main/java/org/springframework/integration/handler/support/MessagingMethodInvokerHelper.java

```diff
@@ -1093,13 +1094,18 @@ void setInvocableHandlerMethod(InvocableHandlerMethod newInvocableHandlerMethod)
 			this.invocableHandlerMethod = newInvocableHandlerMethod;
 		}
 
+		@Nullable
 		public Object invoke(ParametersWrapper parameters) {
 			Message<?> message = parameters.getMessage();
 			if (this.canProcessMessageList) {
 				message = new MutableMessage<>(parameters.getMessages(), parameters.getHeaders());
 			}
 			try {
-				return this.invocableHandlerMethod.invoke(message);
+				Object result = this.invocableHandlerMethod.invoke(message);
+				if (result != null && result.getClass().getName().equals("kotlin.Unit")) {
```

Comment:
Can we not capture this condition as a final `boolean` by checking the return type in a CTOR (to avoid running this code on every non-null return)?

Reply:
Well, we can avoid something indeed, when a return type is `void`  or `Unit`. However in this case the return type is `Object` so end-user can return `null` ()or just don't have `return` in the function lambda and we don't produce a reply message.
In this case Kotlin for non-return in function lambda produces for us that `Unit` which is not a `null` apparently.
So, we produce some weird reply message.
Therefore even if we have some final checks for return type, we still would have some logic to check the returned value to be sure that we don't produce a reply message with `Unit` payload.

To summarize: when have `void` return this `result` is always going to be `null`, therefore we don't go to check a class name for `Unit`. If the value not null, it already doesn't matter what `void` check we have in advance: we still have to check this `result` for `Unit` type.

Does it make sense?

Reply:
Could we at least do something like...
```java
if (result != null && this.isKotlinLambda && result.getClass().getName().equals("kotlin.Unit")) {
```

?

This `result.getClass().getName().equals("kotlin.Unit")` just seems like a lot of overhead for this one corner case and not good for the 99.9% of calls.

Reply:
Ah! OK. Makes sense.
I'll think what we can do...
Thanks for the pointer!